File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        # will use dfs approach
        root_val = root.val
        count = 0
        
        def dfs(node, prev):
            nonlocal count
            
            if node is None:
                return 0
            
            if node.val >= prev.val and node.val >= root_val:
                count += 1
            
            left_branch = dfs(node.left, node)
            right_branch = dfs(node.right, node)

            return count
        

        return dfs(root, root)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> output;
        vector<int> result = treeList(root, output);
        return result[k-1];
    }

    vector<int> treeList(TreeNode* root, vector<int>& result) {
        if (!root) return result;

        if (root->left && root->left->left == nullptr) {
            result.push_back(root->left->val);
        }
        result.push_back(root);
        if (root->right && root->right->right == nullptr) {
            result.push_back(root->right->val);
        }
        
        return treeList(root->left, result) && treeList(root->right, result);
    }
};

